### PR TITLE
DEMO: [rtsan] Support suppressions list

### PR DESCRIPTION
**Pull request by cjappl**

### CODE REVIEW

File: compiler-rt/lib/rtsan/rtsan_checks.inc [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+//===----------------------------------------------------------------------===//
+#ifndef RTSAN_CHECK
+#error "Define RTSAN_CHECK prior to including this file!"
+#endif
+
+// RTSAN_CHECK(Name, SummaryKind)
+// SummaryKind should be a string literal.
+
+RTSAN_CHECK(FunctionCall, "function-call")
+RTSAN_CHECK(Interceptor, "interceptor")
```
Comment (cjappl) [inline compiler-rt/lib/rtsan/rtsan_checks.inc:20]: Do we like the names of these? What else would we call them

FunctionCall is "does this call exist anywhere in the stack"


Interceptor right now disables checks at the lowest level for a blocking function or an intercepted call by name (`malloc` or `my_call_with_attr_blocking`)

* Do we want to have two different disabling for each of them?
    * If so, what are their names and strings?
* Do we even want people to be able to disable [[blocking]] calls?

File: compiler-rt/lib/rtsan/rtsan.cpp [hunk 2]
```diff
@@ -46,38 +46,30 @@ static InitializationState GetInitializationState() {
       atomic_load(&rtsan_initialized, memory_order_acquire));
 }
 
-static auto OnViolationAction(DiagnosticsInfo info) {
-  return [info]() {
-    IncrementTotalErrorCount();
+static auto OnViolationAction(const BufferedStackTrace &stack,
+                              const DiagnosticsInfo &info) {
```
Comment (cjappl) [inline compiler-rt/lib/rtsan/rtsan.cpp:50]: Note that now, when a violation occurs, ExpectNotRealtime passes in the stack and diagnostics info for more processing.

This function is probably best shown with whitespace turned off, I mostly just extracted things out of the lambda

File: compiler-rt/lib/rtsan/rtsan_assertions.h [hunk 3]
```diff
@@ -14,16 +14,48 @@
...
+public:
+  [[nodiscard]] explicit ScopedBypass(Context &context) : context_(context) {
+    context_.BypassPush();
+  }
+
+  ~ScopedBypass() { context_.BypassPop(); }
+
+private:
+  Context &context_;
+};
```
Comment (cjappl) [inline compiler-rt/lib/rtsan/rtsan_assertions.h:34]: If we go this direction, I will split this into it's own review, likely putting this in rtsan_context.h

**Comment by cjappl** [comment]
Overall, in this V2 I had two initial ways to go about it:

1. This way, preserving the idea of the "OnViolationAction" being separate from the question "Is this a violation" (achieved in ExpectNotRealtime, now weighing suppressions into account)
2. Moving back to the "old style" ExpectNotRealtime - a long imperative function call with both detection of a violation, and doing something about a violation.

The second is easy to imagine, so I opted to "show and tell" the first one. Interested to hear if you have a preference on one versus the other. 

I'm wondering if there is any benefit to keeping this separate like this, or if we should compress into a big imperative function. (This is not intended to be leading, I find myself on the fence either way and would value your input)

File: compiler-rt/lib/rtsan/rtsan_checks.inc (hunk 1 above)
Comment (davidtrevelyan) [reply compiler-rt/lib/rtsan/rtsan_checks.inc:20]: Yeah good questions. Here's my initial thoughts:

_Q: Should we have different disabling keys for blocking, intercepted and stack functions?_

A:  I believe so, yes, for all three. I like to think we can improve on the names to make them more immediately understandable, even to people without a mental model of how the sanitizer works, but I'm struggling with `interceptor`. Here's my first iteration on the naming, but I don't think they're perfect at all, so keen to iterate more on them with you:

- "improve" `function-call` to something more like `call-stack-match`
- `interceptor` -> `intercepted`
- add `blocking`

_Q: Do we even want people to suppress `[[blocking]]` calls?_

A: I think it could still be helpful, yes - I can imagine a situation where you're using a third-party library or re-using multi-threaded code in a single-threaded context (an noncontended spinlock here would always immediately grab a lock - the same reasoning here as it would apply to a normal mutex imho). I also worry that not suppressing them could potentially even be more confusing than otherwise. 

Another question for you:

_Q: Should we allow regex for `intercepted` and `blocking`? I can imagine a user might want to just suppress all `pthread_.*`, for example..._

File: compiler-rt/lib/rtsan/rtsan_assertions.h (hunk 3 above)
Comment (davidtrevelyan) [reply compiler-rt/lib/rtsan/rtsan_assertions.h:34]: Could we do without this and use `__rtsan::ScopedDisabler` instead?

File: compiler-rt/lib/rtsan/rtsan_checks.inc (hunk 1 above)
Comment (davidtrevelyan) [reply compiler-rt/lib/rtsan/rtsan_checks.inc:20]: Having said that, regarding naming, if there is precedent for these sorts of names in other sanitizers, I appreciate there might be a good argument for matching those...

File: compiler-rt/lib/rtsan/rtsan_suppressions.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  if (!suppression_ctx->HasSuppressionType(function_call_type))
+    return false;
+
+  Symbolizer *symbolizer = Symbolizer::GetOrInit();
+  Suppression *s;
+
+  for (uptr i = 0; i < stack.size && stack.trace[i]; i++) {
+    uptr addr = stack.trace[i];
+
+    if (suppression_ctx->HasSuppressionType(function_call_type)) {
```
Comment (davidtrevelyan) [inline compiler-rt/lib/rtsan/rtsan_suppressions.cpp:76]: Seems like there's quite a lot of code here that looks very similar to what's in asan's suppression implementation. Do you think it's worth abstracting anything in common into shared utility functions, or are they different enough to warrant their own implementations?

**Review by davidtrevelyan — CHANGES_REQUESTED** [review]

**Comment by davidtrevelyan** [comment]
> Overall, in this V2 I had two initial ways to go about it:
> 
> 1. This way, preserving the idea of the "OnViolationAction" being separate from the question "Is this a violation" (achieved in ExpectNotRealtime, now weighing suppressions into account)
> 2. Moving back to the "old style" ExpectNotRealtime - a long imperative function call with both detection of a violation, and doing something about a violation.
> 
> The second is easy to imagine, so I opted to "show and tell" the first one. Interested to hear if you have a preference on one versus the other.
> 
> I'm wondering if there is any benefit to keeping this separate like this, or if we should compress into a big imperative function. (This is not intended to be leading, I find myself on the fence either way and would value your input)

I personally consider option 1. to have better separation of concerns and to be far easier to test (no need to EXPECT_DEATH, which forks the process, any more with a fake OnViolationAction), so I prefer that one at the moment. Flipping the question on its head - what do you see as the benefit of having the "old style" imperative function here?

File: compiler-rt/lib/rtsan/rtsan_assertions.h (hunk 3 above)
Comment (cjappl) [reply compiler-rt/lib/rtsan/rtsan_assertions.h:34]: Hmm, I'm not sure if we have access to that construct here, but I'll investigate

**Comment by cjappl** [comment]
> I personally consider option 1. to have better separation of concerns and to be far easier to test (no need to EXPECT_DEATH, which forks the process, any more with a fake OnViolationAction), so I prefer that one at the moment. Flipping the question on its head - what do you see as the benefit of having the "old style" imperative function here?

This implementation is good with me 👍 I think it is the better of the two regarding separation of concerns etc

File: compiler-rt/lib/rtsan/rtsan_checks.inc (hunk 1 above)
Comment (cjappl) [reply compiler-rt/lib/rtsan/rtsan_checks.inc:20]: > Q: Should we have different disabling keys for blocking, intercepted and stack functions?
> A: I believe so, yes, for all three.

Sounds good, I'm in favor of this. I think the only reason to avoid any of it would be implementation/runtime complexity, but both of those are very tame. I think more atomic is better

> "improve" function-call to something more like call-stack-match
> interceptor -> intercepted
> add blocking

I like these generally, I'll give it some thought and we can continue hashing it out on the end PR

> _Q: Do we even want people to suppress [[blocking]] calls?
> A: I think it could still be helpful, yes

👍 

> Another question for you:
> Q: Should we allow regex for intercepted and blocking? I can imagine a user might want to just suppress all pthread_.*, for example...

Perhaps, but I would encourage punting on this until later on. The regex engine definitely will slow down the string matching. Before we go deeper with that I'd want to let the "basic" version prove to be inadequate for folks.

Basically "can we prove that we need it" before we implement something more complex. I think we are providing a lot of flexibility to the users already

**Comment by davidtrevelyan** [comment]
> > I personally consider option 1. to have better separation of concerns and to be far easier to test (no need to EXPECT_DEATH, which forks the process, any more with a fake OnViolationAction), so I prefer that one at the moment. Flipping the question on its head - what do you see as the benefit of having the "old style" imperative function here?
> 
> This implementation is good with me 👍 I think it is the better of the two regarding separation of concerns etc

Nice - agreed 👍

File: compiler-rt/lib/rtsan/rtsan_suppressions.cpp (hunk 4 above)
Comment (cjappl) [reply compiler-rt/lib/rtsan/rtsan_suppressions.cpp:76]: Good idea. I would recommend investigating this in a follow up review. I don't want to abstract too soon and muddy up this change in isolation.

This could be a good refactor follow-on.

**Review by cjappl — COMMENTED** [review]
Thanks for the thoughts, greatly appreciated.

I am going to close this review, and put a review up against LLVM/main where we can continue discussing etc with your feedback incorporated

File: compiler-rt/lib/rtsan/rtsan_assertions.h (hunk 3 above)
Comment (cjappl) [reply compiler-rt/lib/rtsan/rtsan_assertions.h:34]: The answer is that no compiler-rt libs include these headers, except in tests. If you do a grep such as:

```
> rg "ubsan_interface.h" --glob "!*test*"
llvm/utils/gn/secondary/compiler-rt/include/BUILD.gn
25:    "sanitizer/ubsan_interface.h",

compiler-rt/include/sanitizer/ubsan_interface.h
1://===-- sanitizer/ubsan_interface.h -----------------------------*- C++ -*-===//
13:#ifndef SANITIZER_UBSAN_INTERFACE_H
14:#define SANITIZER_UBSAN_INTERFACE_H
32:#endif // SANITIZER_UBSAN_INTERFACE_H

compiler-rt/include/CMakeLists.txt
```

This shows up blank for all interface files. Based on this (for better or worse) we are left with re-implementing anything in our "top level header" down here in the depths of the impl.

We can theorize how to improve this (perhaps for us and all sanitizers), but I'm going to keep this as is for now

File: compiler-rt/lib/rtsan/rtsan_checks.inc (hunk 1 above)
Comment (cjappl) [reply compiler-rt/lib/rtsan/rtsan_checks.inc:20]: One correction here (of myself):

> Another question for you:
> Q: Should we allow regex for intercepted and blocking? I can imagine a user might want to just suppress all pthread_.*, for example...

So by default (and without me knowing) we ARE using the regex engine for function matches. This is the function we use:

```cpp
bool SuppressionContext::Match(const char *str, const char *type,
                               Suppression **s) {
  can_parse_ = false;
  if (!HasSuppressionType(type))
    return false;
  for (uptr i = 0; i < suppressions_.size(); i++) {
    Suppression &cur = suppressions_[i];
    if (0 == internal_strcmp(cur.type, type) && TemplateMatch(cur.templ, str)) {
      *s = &cur;
      return true;
    }
  }
  return false;
}

```

So early out for strcmp, then TemplateMatch, which is "regex light". It looks like it has a runtime of O(N), so it isn't more costly than just strcmp. We do both strcmp and TemplateMatch per function specified per type, so O(N_strlen*M_num_interceptor_suppressions).

Something to be aware of, that I don't think we will be able to get around - 

If the user is using suppressions, the more suppressions they use the more that run-time is affected. We need to be clear that using a ScopedDisabler is the first tool they should reach for if at all possible, which has an O(1) runtime cost. 

The hierarchy of speed:
* ScopedDisabler (easy, early out, never compare strings, never unwind the stack) - O(1)
* Function name suppressions, blocking and interceptors - O(N_strlen * M_suppressions_in_type)
* Stack suppressions - O(N_strlen * M_suppressions_in_type * L_stack_frame_count)

(I think)
